### ISSUE TITLE
`fs-storage`: Port Monoid-based conflicts resolution from Kotlin

### ISSUE THREAD

**Issue by kirillt** [ARK-Builders/ark-rust#19]
Port [Monoid structure](https://github.com/ARK-Builders/arklib-android/blob/main/lib/src/main/java/dev/arkbuilders/arklib/data/storage/Monoid.kt) used during conflicts resolution. It's naive approach but it's simpler than CRDT.

**Comment by twitu** [comment]
https://github.com/ARK-Builders/arklib-android/blob/main/lib/src/main/java/dev/arkbuilders/arklib/data/storage/Monoid.kt#L3

The kotlin monoid implementation is functionally equivalent to what modern CRDT libraries are offering. While we can go for a custom implementation it is error prone and require extra dev effort for implementing and maintaining. I can look into existing CRDT rust implementations for a feature set that suits us.

**Comment by twitu** [comment]
Rust has two popular libraries for CRDTs - [crdts](https://docs.rs/crdts/latest/crdts/index.html) and [automerge](https://docs.rs/automerge/latest/automerge/). It appears that automerge is the more popular, well-maintained and well-featured library.

Now for the actual structures I looked at the actual structures that implement monoids in the Kotlin library

| Class | Structure |
| -- | -- |
| [Tags](https://github.com/ARK-Builders/arklib-android/blob/main/lib/src/main/java/dev/arkbuilders/arklib/user/tags/Tags.kt#L18) | Set |
| [Properties](https://github.com/ARK-Builders/arklib-android/blob/main/lib/src/main/java/dev/arkbuilders/arklib/user/properties/Properties.kt#L13) | Set |
| [Score](https://github.com/ARK-Builders/arklib-android/blob/main/lib/src/main/java/dev/arkbuilders/arklib/user/score/Score.kt#L9) | Int |

So it appears that for the current structures we don't need to pull in a CRDT implementation. In fact `HashSet` already implement a union function, in fact Kotlin too uses the standard library implementation for union. So only a special function for integers is needed. CRDTs can be considered later when we need to add structures that need more powerful combine semantics.

**Comment by kirillt** [comment]
@twitu yes, if we can use a better structure we could throw Monoids away. You are right that `HashSet` is a monoid itself, I need to refresh my memory about how exactly Monoids been used in Kotlin implementation.

But I remember that there was a problem with Monoid-like structures that we cannot remove data:
1. User has phone **P** and laptop **L** connected via Syncthing, so storages are synced as plain files.
2. Both **P** and **L** run the same app, listening the storages on their disks.
3. User adds entry **X** into the app on **P**, it is persisted, synced to **L**'s disk and loaded into **L**'s instance of the app.
4. Now, the user deletes **X** from the app on **P**, we remove it from **P**'s disk, app on **L** detects the change but overwrites the storage with older version because _union(S+X, S)=S+X for any storage S and entry X_.

Can we solve this problem using CRDTs?

**Comment by kirillt** [comment]
Btw, am I right that [automerge](https://docs.rs/automerge/latest/automerge/) has Rust core and JS wrapper? I just see that there are much more JS code in their repo.

**Comment by twitu** [comment]
Monoids cannot handle the delete case directly. There are ways to use monoids for delete by adding soft delete flags for records. That is a pretty good solution for most use cases. Monoids will be a good solution for the uses cases (HashSet/Int) that we have in the Kotlin app.

Yes automerge has a rust core and JS bindings for WASM. I guess they've strongly positioned for use with client side web apps so it has some JS to make it ergonomic to use on the client side. It's still a pretty well featured rust library we can use if needed.

My concern however is that it uses it's own data structures for Maps and Lists and the likes. So there will be a conversion overhead from the standard library BTreeMap/HashMap/HashSet to automerge's data structures so that different versions can be merged. I would put off pulling in such a complex dependencies until we really need it.

**Comment by kirillt** [comment]
We'll be able to define a single "sync with disk" function, similarly to how it was in [BaseStorage.kt](https://github.com/ARK-Builders/arklib-android/blob/main/lib/src/main/java/dev/arkbuilders/arklib/data/storage/BaseStorage.kt), after the conflicts resolution is implemented.

**Comment by kirillt** [comment]
Ishan's idea: no need to implement CRDT if we just do "soft-delete" instead of regular deletion. Adding second Monoid instance would help us in syncing deletions between devices.

**Comment by kirillt** [comment]
Merged:
- #26

We also need:
- implement a function for updating the storage with the disk (extract new mapping from fs, call `merge_from`)
- discuss writing strategies:
    - how often we persist the updated model?
    - do we persist the updated model immediately after the sync?
    - do we allow manual syncing?
    - we might need something like a `PersistenceStrategy`:
        - struct with boolean fields corresponding to these questions
        - trait with hooks handling sync events, this one could also be used for other scenarios

**Comment by kirillt** [comment]
Links for the research of CRDTs:
- Rust
  - "Diamond Types"
    - https://docs.rs/diamond-types/latest/diamond_types/
    - https://github.com/josephg/diamond-types/blob/master/INTERNALS.md
    - https://braid.org/meeting-14
  - https://github.com/y-crdt/y-crdt?tab=readme-ov-file
    - https://yjs.dev/
  - https://braid.org/
    - https://braid.org/demo/interact
- JavaScript:
  - https://github.com/streamich/json-joy

**Comment by kirillt** [comment]
Basing on [this article](https://kalele.io/summary-of-crdts/), it seems that we can use the following data types:
- `PN-counter`s for scores
- `U-set`s for tags
- `U-set`s for properties
- `U-set`s for the storage tables themselves

It seems that `U-set` can be used for maps, not only sets, by interpreting a map as a set of entries.

@tareknaser @twitu

**Comment by twitu** [comment]
One concern I see with `U-set` is that it won't preserve the order of insertion property of `BTreeMap`s. In #63 we implemented something very similar to a last writer wins (LWW) CRDT by merging the BTreeMap values based on the most recent timestamp.

The main question to consider is how we intend to use insertion order?

Perhaps, if we can relax the requirement for maintaining insertion order we can consider using the `U-set` or maybe the `LWW-Element-Set` which records the timestamp for each element which can be used as an approximation for insertion order.
